server/src-lib/Hasura/RQL/DDL/RemoteSchema/Permission.hs

```diff
@@ -0,0 +1,966 @@
+
+
+import qualified Data.HashMap.Strict                   as Map
+import qualified Language.GraphQL.Draft.Syntax         as G
+import qualified Data.HashSet                          as S
+import qualified Data.List.NonEmpty                    as NE
+import qualified Data.Text                             as T
+
+import           Hasura.Session
+
+data FieldDefinitionType
+  = ObjectField
+  | InterfaceField
+  | EnumField
+  deriving (Show, Eq)
+
+instance ToTxt FieldDefinitionType where
+  toTxt = \case
+    ObjectField    -> "Object"
+    InterfaceField -> "Interface"
+    EnumField      -> "Enum"
+
+data ArgumentDefinitionType
+  = InputObjectArgument
+  | DirectiveArgument
+  deriving (Show, Eq)
+
+instance ToTxt ArgumentDefinitionType where
+  toTxt = \case
+    InputObjectArgument -> "Input object"
+    DirectiveArgument   -> "Directive"
+
+data PresetInputTypeInfo
+  = PresetScalar
+  | PresetEnum ![G.EnumValue]
+  | PresetInputObject ![G.InputValueDefinition]
+  deriving (Show, Eq, Generic, Ord)
+
+data GraphQLType
+  = Enum
+  | InputObject
+  | Object
+  | Interface
+  | Union
+  | Scalar
+  | Directive
+  | Field !FieldDefinitionType
+  | Argument !ArgumentDefinitionType
+  deriving (Show, Eq)
+
+instance ToTxt GraphQLType where
+  toTxt = \case
+    Enum                         -> "Enum"
+    InputObject                  -> "Input object"
+    Object                       -> "Object"
+    Interface                    -> "Interface"
+    Union                        -> "Union"
+    Scalar                       -> "Scalar"
+    Directive                    -> "Directive"
+    Field ObjectField            -> "Object field"
+    Field InterfaceField         -> "Interface field"
+    Field EnumField              -> "Enum field"
+    Argument InputObjectArgument -> "Input object argument"
+    Argument DirectiveArgument   -> "Directive Argument"
+
+data RoleBasedSchemaValidationError
+  = NonMatchingType !G.Name !GraphQLType !G.GType !G.GType
+  -- ^ error to indicate that a type provided by the user
+  -- differs from the corresponding type defined in the upstream
+  -- remote schema
+  | FieldDoesNotExist !GraphQLType !G.Name
+  -- ^ error to indicate when a field doesn't exist
+  -- in the upstream remote schema
+  | NonMatchingDefaultValue !G.Name !G.Name !(Maybe (G.Value Void)) !(Maybe (G.Value Void))
+  -- ^ error to indicate when the default value of an argument
+  -- differs from the default value of the corresponding argument
+  | NonExistingInputArgument !G.Name !G.Name
+  -- ^ error to indicate when a given input argument doesn't exist
+  -- in the corresponding upstream input object
+  | MissingNonNullableArguments !G.Name !(NonEmpty G.Name)
+  | NonExistingDirectiveArgument !G.Name !GraphQLType !G.Name !(NonEmpty G.Name)
+  -- ^ error to indicate when a given directive argument
+  -- doesn't exist in the corresponding upstream directive
+  | NonExistingField !(FieldDefinitionType, G.Name) !G.Name
```

Comment:
Now, that I re-visit the code, the `FieldDoesNotExist` is a terrible name 😓 , it should actually be `TypeDoesNotExist` because the error is thrown when a user provided type definition doesn't exist in the upstream remote schema.

I have updated the code, the same way.